File: handy-ts/quiz/q5_longest_substring_without_repeating_characters_medium.ts
/**
 * Q5) Longest Substring Without Repeating Characters
 *
 * AI-BEST:
 * Sliding window + last seen index map.
 * Time: O(n), Space: O(k)
 *
 * AI-EASY:
 * Maintain current window string with indexOf/slicing.
 * Time: can degrade to O(n^2), Space: moderate
 */

function longestSubstringBest(s: string): string {
  let left = 0;
  let bestStart = 0;
  let bestLen = 0;
  const lastSeen = new Map<string, number>();

  for (let right = 0; right < s.length; right++) {
    const ch = s[right];
    if (lastSeen.has(ch) && (lastSeen.get(ch) ?? -1) >= left) {
      left = (lastSeen.get(ch) ?? 0) + 1;
    }
    lastSeen.set(ch, right);

    const currentLen = right - left + 1;
    if (currentLen > bestLen) {
      bestLen = currentLen;
      bestStart = left;
    }
  }

  return s.slice(bestStart, bestStart + bestLen);
}

function longestSubstringEasy(s: string): string {
  let currentWindow = "";
  let longestFound = "";

  for (const ch of s) {
    const duplicateIndex = currentWindow.indexOf(ch);
    if (duplicateIndex >= 0) {
      currentWindow = currentWindow.slice(duplicateIndex + 1) + ch;
    } else {
      currentWindow += ch;
    }
    if (currentWindow.length > longestFound.length) {
      longestFound = currentWindow;
    }
  }
  return longestFound;
}

function runDemo(): void {
  console.log("Q5: Longest Substring Without Repeating Characters");
  const samples = ["abcdbcabcbb", "bbbbb", "pwwkew", "", "au"];
  for (const s of samples) {
    console.log(`Input: "${s}"`);
    console.log("  BEST:", longestSubstringBest(s));
    console.log("  EASY:", longestSubstringEasy(s));
  }
}

runDemo();

export {};

File: handy-java/quiz/q5_longest_substring_without_repeating_characters_medium.java
import java.util.HashMap;
import java.util.Map;

/**
 * Q5) Longest Substring Without Repeating Characters
 *
 * AI-BEST:
 * Sliding window + last seen index map.
 * Time: O(n), Space: O(k)
 *
 * AI-EASY:
 * String window with indexOf + substring.
 * Time: can degrade to O(n^2), Space: moderate
 */
public class q5_longest_substring_without_repeating_characters_medium {
    static String longestSubstringBest(String s) {
        int left = 0;
        int bestStart = 0;
        int bestLen = 0;
        Map<Character, Integer> lastSeen = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            if (lastSeen.containsKey(ch) && lastSeen.get(ch) >= left) {
                left = lastSeen.get(ch) + 1;
            }
            lastSeen.put(ch, right);

            int currentLen = right - left + 1;
            if (currentLen > bestLen) {
                bestLen = currentLen;
                bestStart = left;
            }
        }

        return s.substring(bestStart, bestStart + bestLen);
    }

    static String longestSubstringEasy(String s) {
        String currentWindow = "";
        String longestFound = "";

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            int duplicateIndex = currentWindow.indexOf(ch);

            if (duplicateIndex >= 0) {
                currentWindow = currentWindow.substring(duplicateIndex + 1) + ch;
            } else {
                currentWindow += ch;
            }

            if (currentWindow.length() > longestFound.length()) {
                longestFound = currentWindow;
            }
        }
        return longestFound;
    }

    public static void main(String[] args) {
        System.out.println("Q5: Longest Substring Without Repeating Characters");
        String[] samples = {"abcdbcabcbb", "bbbbb", "pwwkew", "", "au"};
        for (String s : samples) {
            System.out.println("Input: \"" + s + "\"");
            System.out.println("  BEST: " + longestSubstringBest(s));
            System.out.println("  EASY: " + longestSubstringEasy(s));
        }
    }
}


